### ISSUE TITLE
Preset disable feedback and improvements

### ISSUE THREAD

**Issue by antero111** [antero111/plugin-presets#54]
a0bd64a9fe7fd3e8f26935b5e1819e3b29b629d3 added way for disabling presets. When presets contain some configurations to plugin enabled values e.g. skybox = true, that preset can be disabled with turning the toggle switch off.  Disabling will turn every plugin off that has the enabled = true in the preset. 

With this you can make easy toggles for plugins by simply creating a preset and then adding some configurations to it. Turning the preset off will turn the plugins off and turning on will enable them back. It does not edit plugin settings so if you wan't to edit e.g. skybox colour, it has to be made as before.

I'd like to hear your toughts and some feedback about this feature

**Comment by antero111** [comment]
@Jademalo Would you be interested to test this and give some feedback?

**Comment by Jademalo** [comment]
Seems to be working as I would expect, though I did note a couple of little things.

If a profile is disabled that is also an auto-updating profile, the auto-updating should be set to standby before the profile is disabled, same as if another profile is loaded. Disabling a preset like this can lead to some weird situations, such as;
![image](https://user-images.githubusercontent.com/386846/217981975-d4faaa64-e777-4a8d-856b-5041b0bdeaa4.png)
![image](https://user-images.githubusercontent.com/386846/217982017-3d45d3f5-b762-4a62-985b-ccf7533f20cf.png)

If you have two presets that are the same except one of them has a plugin on/off setting selected, when you select the preset without that selected the other preset will flicker on and off. 
(Actually, I think this is the same bug as I mentioned in #53)

_____

The only other bit of feedback I have is it can feel a bit... Unclear?
I've ended up in a couple of situations when testing where a plugin has been disabled without me noticing, since there's no feedback that it's been disabled instead of a profile just not matching.

I'm wondering if there needs to be a third visual state for the toggle switch associated with this disabled state, since it's pretty easy now to end up in weird situations, but after a good bit of thinking about it I can't really think of a good way to make it work.
Another thought is maybe require this to be specifically enabled on a per-preset basis, but I think that might actually make the vagueness worse.
Another idea could be to automatically and forcibly enable plugin on/off state as a saved option if a preset has any settings selected for a plugin, but that again feels like a less than elegant solution since you're forcing something to be set.

I think ultimately I feel like as currently implemented, this conflicts with how the plugin is meant to work adding an off switch to something that can't entirely be turned off.
If the core design of the plugin was different such as having a base config set and then the presets all being overlays, this would make a lot more sense I think since you'd be able to "disable" a specific preset as a whole. That's ultimately a different plugin entirely, though.

______

After having a think, the way I think I would implement this honestly is to simplify and separate this by having a totally different type of preset that specifically has the job of enabling and disabling plugins. 

Maybe have another right click option on the + icon for "Create new toggle preset", which gives a simplified list that lets you select which plugins you want it to control. Enabling this preset enables the selected plugins, disabling it disables the selected plugins, nothing else. 
This could then have 3 visual states on the toggle - Yellow on for matching/all enabled, Red off for all disabled, and grey off for partially matching (so some are enabled and some are disabled).
There would probably need to be a visual indication that the preset is a toggle preset instead of a standard preset, but I think this would go a long way in terms of getting rid of vagueness and keeping functionality clear.

___

So to sum up my thoughts - It works well as implemented, but for me personally I feel like the current implementation conflicts with the core design of the plugin.

**Comment by antero111** [comment]
Thanks for the feedback. You are right about it feeling a little unclear or "out of the design". This was a little test I wanted to make since having a feature that allows for simple toggling would be usefull most of the basic RL people out there. Not many people needs a complex preset setup for every activity, mostly just a simple toggle for fishing plugin when fishing etc. This would make it possible to easily switch multiple plugins on/off instead of using the settings tab. 

This commit didn't include any ux changes other that the preset disabling since that is something to be discussed more later when the design is finnished. Most importantly you made it clear that it works but its not very intuitive.

I was thinking of some sort of toggle for presets for having "toggler presets" before making this but I wanted to test if this was good as it is but clearly its not. You are right that it conflicts with the base logic of the plugin since you really don't have any means of seeing if preset was manually disabled or it was toggled of because the configs differ. This could be fixed with the third style of toggle or some other ui addition like some text describing a preset is a "toggler". 

The end goal for this is to have a way for easily toggle plugins included in presets for easy switches **in one preset**. The question is how do we implement this. You can do this now with having 2 presets that store different enabled values e.g. for fishing and the switch between them. Maybe you could convert some preset to be a toggler or alternatively you could "stack" 2 presets together by storing configs that get loaded when the preset is enabled and what when disabled. 

One thing is sure that if there is preset disabling (the switch turning off) there needs to be a way to see if preset is disabled or if it does not match your configurations.

**Comment by Jademalo** [comment]
>Or alternatively you could "stack" 2 presets together by storing configs that get loaded when the preset is enabled and what when disabled.

I had originally written a lot more here, but honestly after thinking I think this is the best way to go about it. Link two presets that represent two states into one, allowing an on/off toggle.
This would function essentially like the overlays thing I mentioned, it would allow a lot of flexibility, and it would stay true to the design of the plugin imo.

In my favourite Main/Iron example, those two presets could be collapsed into one called Iron. Enabling it overlays the config necessary for playing an iron, disabling it loads the config for a standard acc.

**Comment by antero111** [comment]
> he only thing I can think is a toggle switch that goes between left, middle, and right, with left and right being yellow and middle being grey

I'd like to keep the simple on/off switch because it matches with the RuneLite settigns and mostly "its part of the design". I tought about having current on off for the configurations matching and one much darker off switch for displaying if a preset is disabled.

How would the preset linking or stacking work ux wise? Right click context menu like item highlights? Drag and drop is out of the question client wise.

If there was a way to stack presets, I'd say there still should be a way to convert one single preset to a toggler (e.g. it converts itself to a stacked preset with on anf off states). Current preset stacking would be just a addition to the current plugin logic and the toggler switch would be a nice addition to that. Making stacked on/off states is somewhat time consuming so for simple toggles, that is needed.

Maybe something like this for displaying the preset has been linked or stacked to some other and then have similiar double arrow icon to swap between the presets. Toggling the preset off would bring the other to the front and the also have the ability to separate the presets.

![kuva](https://user-images.githubusercontent.com/60603897/218249131-71894b6d-5302-4d8c-bf22-a6e4a9c36443.png)

**Comment by Jademalo** [comment]
I've had a go at something, but honestly this feels like a different feature entirely. I'm struggling to think of a good idea without inventing a new UI element for a three state toggle. 

![image](https://user-images.githubusercontent.com/386846/218252588-b974fa79-65df-4d9a-947f-04f66cf4b901.png)

This is the best I can come up with, where the linked profile you select is given a highlighted orange name. Clicking the other name will switch to that profile.

...but then what is the point of this, why not just use two profiles as you can now.
The problem is that the toggle switch is very specifically On/Off in it's design, but we're trying to use it to denote A/B (or C). 

Green and red are used in the standard Runelite UI for the Google account login, the only thing I can think is having the toggle switch be red on the left/off position, green on the right/on position, and grey if neither match. But that definitely feels like a stretch as far as UI conventions go.

I'll have another think, but I'm struggling to come up with any ideas. The problem is that if we've got two profiles stacked/linked, we're using the switch to denote A/B when in all other scenatios in the plugin the switch means matching/not matching.

**Comment by Jademalo** [comment]
> How do we denote that a preset is a "toggler"? 

I think maybe an icon similar to the cloud icon now, next to the name. I don't know what icon to use, but that seems most logical to me. It might be worth having icons for both standard and toggler presets honestly, for consistency sake.

> Do toggler presets include also plugin settigns or just the enabled state? 

In my opinion, I think the right way to do this is to *just* have the enabled state. Make these presets very basic, with only a single function. If you want more complex function you'd use preset pairs anyway, this is specifically a simple easy toggle so the functionality should be simple imo.

> How should they behave when the preset has been enabled and then its configs differ after a while?

Assuming by configs differ we're talking about plugins being inconsistently enabled/disabled, I think the best way to handle togglers would be to not change the toggle switch at all automatically, and only have it change on user input.

If the user toggles it from the off to on state, all plugins associated with it are turned on, regardless of their current state.
If the user toggles it from the on to off state, all plugins associated with it are turned off, regardless of their current state.
If a user got it into a weird state where the preset was set to on but some plugins were off and they wanted to use it to turn all plugins on, they could flip it twice.

It might be worth having a separate indicator for if it's out of "sync", maybe turning the icon that indicates it being a toggle preset grey or something?

> I don't know if this feature is just something I think someone would need and maybe we should think of some real scenario where toggler preset and also normal presets were used. The thing I like about the togglers is that its very easy to use and does what I think most people do... turn plugins on/off between different activities. I don't think many people change plugin settings so often other than red player indicators to wildy etc etc. And also togglers are just one on/off preset so much simpler to understand vs current presets.

Honestly I think you've answered your own question - This is for people who just want a very basic on/off for a set of plugins. Sure there are people like me who will really go deep and use the complex features to their full effect, but this will definitely be useful for some imo.

**Comment by antero111** [comment]
I'll see what I can come up with. Currently wating for RuneLites core profile switcher to be released https://github.com/runelite/runelite/commit/831189459ba51538b239a33b7d632bd2d9e4e8a3. I think that we need to make some changes to this plugin to not do things that the core will do so I'm waiting a little with this to see how that goes.

**Comment by Jademalo** [comment]
Oh wow, interesting to see it finally be tackled in core. Honestly wasn't expecting that
